insights/client/crypto.py

```diff
@@ -0,0 +1,157 @@
+        self._home = tempfile.mkdtemp()
+
+        logger.debug("setting up gpg in the temporary environment")
+        for key in self.keys:
+            self._run(["--import", key])
+
+            if not self.ok:
+                logger.debug("failed to import key {}".format(key))
+                return False
+
+        return True
+
+    def _cleanup(self):
+        shutil.rmtree(self._home)
+
+    def _run(self, command):
+        self._raw_command = ["/usr/bin/gpg", "--homedir", self._home] + command
+        process = subprocess.Popen(
+            self._raw_command,
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE,
+        )
+        stdout, stderr = process.communicate()
+
+        self.ok = process.returncode == 0
+        self.return_code = process.returncode
+
+        if self.ok:
+            logger.debug("gpg command {command}: ok".format(command=command))
+        else:
+            logger.debug(
+                "gpg command {command} returned status code {code}".format(
+                    command=command, code=self.return_code
+                )
+            )
+
+        self.stdout = stdout.decode("utf-8")
+        self.stderr = stderr.decode("utf-8")
+
+        if self.return_code > 0 and self.stdout:
+            logger.debug("stdout is not empty:\n%s", self.stdout.strip())
+        if self.return_code > 0 and self.stderr:
+            logger.debug("stderr is not empty:\n%s", self.stderr.strip())
+
+    def evaluate(self):
+        try:
+            ok = self._setup()
+            if not ok:
+                return
+
+            logger.debug("running gpg in the temporary environment")
+            self._run(self.command)
+        finally:
+            self._cleanup()
+
+
+def verify_gpg_signed_file(file, signature, keys):
+    """
+    Verify a file that was signed using GPG.
+
+    Parameters:
+        file (str): A path to the signed file.
+        signature (str): A path to the detached signature.
+        keys (list(str)):
+            List of paths to GPG public keys on the filesystem to check against.
+
+    Returns (GPG): Evaluated GPG command.
+    """
+    if not os.path.isfile(file):
+        logger.debug(
+            "cannot verify signature of '{file}', file does not exist".format(
+                file=file
+            )
+        )
+        raise OSError(errno.ENOENT, "File '{file}' not found".format(file=file))
+
+    if not os.path.isfile(signature):
+        logger.debug((
+            "cannot verify signature of '{file}', "
+            "signature '{signature}' does not exist"
+        ).format(file=file, signature=signature))
+        raise OSError(
+            errno.ENOENT,
+            "Signature '{sig}' of file '{file}' not found.".format(
+                sig=signature, file=file
+            )
+        )
+
+    gpg = GPG(["--verify", signature, file], keys)
+    logger.debug("starting gpg verification process for '{file}'".format(file=file))
+
+    gpg.evaluate()
+
+    if gpg.ok:
```

Comment:
A code cleanliness note: `subprocess.Popen` follows the model that you are using here where you do some setup, run a method, and then read the attributes from the object but it's not a very good API idiom to emulate because it's hard to get right, not very flexible, and can lead to bugs down the line.  As an example, if you use `subprocess.Popen.communicate()` twice, it will only run the program once.  The second run will just return cached return information.  On the other hand, your implementation will actually run the program again.  This is a subtle difference that can lead to bugs if future programmers are not careful because the attributes that hold the return values could be modified in that case.

What you're using this for could be done by a function which just returns the values instead of setting them as attributes.  I think the reason you used an object was to separate the generic use of  `--homedir` from the specific command that you want to use.  Some better idioms that python provides for that:

* The main `gpg_run(["--verify", signature, file], key)` simply calls some helper functions for the generic parts.
* A Callable object where you can customize some behaviour of the function when you instantiate the Callable and then it acts as a normal function when you call it.  This is probably closest to the model you have now.  The big difference is that it leads you to return the values the function generates instead of placing them into object attributes.
* A decorator:  You can encapsulate the generic parts into a decorator which sets up the homedir and adds it to the command passed in, runs the function with the modified command, and then cleans up afterwards.

I just thought I'd point this out since copying `subprocess.Popen's` API model was a mistake I sometimes made until I thought about why it wasn't common to see that idiom in other places and realized what the problems were.

Reply:
Yeah. I seemed like an OK idea at first, but looking back a day later it already seemed confusing :) I've split it into two separate objects. I didn't do it at first because it doesn't seem like a common things that's done in this repo, but it improved readability in a significant way.